staging/src/k8s.io/kubectl/pkg/kuberc/kuberc.go

```diff
@@ -0,0 +1,408 @@
+	kuberc "k8s.io/kubectl/pkg/config/install"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/homedir"
+)
+
+const RecommendedKubeRCFileName = "kuberc"
+
+var (
+	RecommendedConfigDir  = filepath.Join(homedir.HomeDir(), clientcmd.RecommendedHomeDir)
+	RecommendedKubeRCFile = filepath.Join(RecommendedConfigDir, RecommendedKubeRCFileName)
+
+	aliasNameRegex = regexp.MustCompile("^[a-zA-Z]+$")
+	shortHandRegex = regexp.MustCompile("^-[a-zA-Z]([-a-zA-Z]*)$")
+
+	scheme = runtime.NewScheme()
+	codecs = serializer.NewCodecFactory(scheme)
+)
+
+func init() {
+	kuberc.Install(scheme)
+}
+
+// PreferencesHandler is responsible for setting default flags
+// arguments based on user's kuberc configuration.
+type PreferencesHandler interface {
+	AddFlags(flags *pflag.FlagSet)
+	Apply(rootCmd *cobra.Command, args []string, errOut io.Writer) ([]string, error)
+}
+
+// Preferences stores the kuberc file coming either from environment variable
+// or file from set in flag or the default kuberc path.
+type Preferences struct {
+	getPreferencesFunc func(kuberc string) (*config.Preference, error)
+
+	aliases map[string]struct{}
+}
+
+// NewPreferences returns initialized Prefrences object.
+func NewPreferences() PreferencesHandler {
+	return &Preferences{
+		getPreferencesFunc: DefaultGetPreferences,
+		aliases:            make(map[string]struct{}),
+	}
+}
+
+type aliasing struct {
+	args    []string
+	flags   []config.CommandOverrideFlag
+	command *cobra.Command
+}
+
+// AddFlags adds kuberc related flags into the command.
+func (p *Preferences) AddFlags(flags *pflag.FlagSet) {
+	flags.String("kuberc", "", "Path to the kuberc file to use for preferences.")
+}
+
+// Apply firstly applies the aliases in the preferences file and secondly overrides
+// the default values of flags.
+func (p *Preferences) Apply(rootCmd *cobra.Command, args []string, errOut io.Writer) ([]string, error) {
+	if len(args) <= 1 {
+		return args, nil
+	}
+
+	kubercPath := getExplicitKuberc(args)
+	kuberc, err := p.getPreferencesFunc(kubercPath)
+	if err != nil {
+		return args, fmt.Errorf("kuberc error %w", err)
+	}
+
+	if kuberc == nil {
+		return args, nil
+	}
+
+	err = validate(kuberc)
+	if err != nil {
+		return args, err
+	}
+
+	args, err = p.applyAliases(rootCmd, kuberc, args, errOut)
+	if err != nil {
+		return args, err
+	}
+	err = p.applyOverrides(rootCmd, kuberc, args, errOut)
+	if err != nil {
+		return args, err
+	}
+	return args, nil
+}
+
+// applyOverrides finds the command and sets the defaulted flag values in kuberc.
+func (p *Preferences) applyOverrides(rootCmd *cobra.Command, kuberc *config.Preference, args []string, errOut io.Writer) error {
+	args = args[1:]
+	cmd, _, err := rootCmd.Find(args)
+	if err != nil {
+		fmt.Fprintf(errOut, "Warning: command not found %v\n", err)
```

Comment:
not sure we want to error on this ... will this output an error when invoking a kubectl plugin command?

Reply:
yes, this would be a problem against plugins. Removed.